File: coding-test/programmers/src/kr/co/programmers/kakao_blind_recruitment_2018/Cache.java
package kr.co.programmers.kakao_blind_recruitment_2018;

import java.util.ArrayList;

public class Cache {

    public static void main(String[] args) {
        int cacheSize = 3;
        String[] cities = {"Jeju", "Pangyo", "Seoul", "NewYork", "LA", "Jeju", "Pangyo", "Seoul", "NewYork", "LA"};
        System.out.println(new Solution().solution(cacheSize, cities));
    }

    static class Solution {
        public int solution(int cacheSize, String[] cities) {
            int answer = 0;

            ArrayList<String> cache = new ArrayList<>();
            for(int i = 0; i < cities.length; i++){
                String element = cities[i].toLowerCase();
                if(cache.contains(element)){
                    answer += 1;
                    cache.remove(element);
                    cache.add(element);
                    continue;
                }

                answer += 5;
                cache.add(element);
                if(cache.size() > cacheSize){
                    cache.remove(0);
                }
            }

            return answer;
        }
    }
}

File: practice-coding-test/javascript/src/kakao-blind-recruitment-2018/cache.js
// @ts-check

/**
 * @param {number} cacheSize
 * @param {string[]} cities
 * @returns {number}
 */
function solution(cacheSize, cities) {
  const cache = []
  return cacheSize === 0
    ? 5 * cities.length
    : cities.reduce((time, city) => {
        const lowerCity = city.toLowerCase()
        const index = cache.indexOf(lowerCity)
        let hit = true
        if (index === -1) {
          hit = false
          if (cache.length >= cacheSize) {
            cache.splice(0, 1)
          }
        } else {
          cache.splice(index, 1)
        }

        cache.push(lowerCity)
        return hit ? time + 1 : time + 5
      }, 0)
}

const cacheSize = 3
const cities = [
  'Jeju',
  'Pangyo',
  'Seoul',
  'NewYork',
  'LA',
  'Jeju',
  'Pangyo',
  'Seoul',
  'NewYork',
  'LA',
]

console.log(solution(cacheSize, cities))
